sql/2022/sustainability/page_bytes_per_type.sql

```diff
@@ -0,0 +1,39 @@
+#standardSQL
+# The distribution of page weight by resource type and client.
+
+# Declare variables to calculate the carbon emissions of one byte
+# Source: https://sustainablewebdesign.org/calculating-digital-emissions/
+# The implementation below does not make the assumptions about returning visitors or caching that are present in the Sustainable Web Design model.
+
+DECLARE kw_per_GB NUMERIC DEFAULT 0.81;
+DECLARE global_grid_intensity NUMERIC DEFAULT 442;
+
+
+SELECT
+  percentile,
+  _TABLE_SUFFIX AS client,
+  APPROX_QUANTILES(bytesTotal / 1024, 1000)[OFFSET(percentile * 10)] AS total_kbytes,
+  APPROX_QUANTILES((bytesTotal / 1024 / 1024 / 1024) * kw_per_GB * global_grid_intensity)[OFFSET(percentile * 10)] AS total_emissions,
+  APPROX_QUANTILES(bytesHtml / 1024, 1000)[OFFSET(percentile * 10)] AS html_kbytes,
+  APPROX_QUANTILES((bytesHtml / 1024 / 1024 / 1024) * kw_per_GB * global_grid_intensity)[OFFSET(percentile * 10)] AS html_emissions,
+  APPROX_QUANTILES(bytesJS / 1024, 1000)[OFFSET(percentile * 10)] AS js_kbytes,
+  APPROX_QUANTILES((bytesJS / 1024 / 1024 / 1024) * kw_per_GB * global_grid_intensity)[OFFSET(percentile * 10)] AS js_emissions,
+  APPROX_QUANTILES(bytesCSS / 1024, 1000)[OFFSET(percentile * 10)] AS css_kbytes,
+  APPROX_QUANTILES((bytesCSS / 1024 / 1024 / 1024) * kw_per_GB * global_grid_intensity)[OFFSET(percentile * 10)] AS css_emissions,
+  APPROX_QUANTILES(bytesImg / 1024, 1000)[OFFSET(percentile * 10)] AS img_kbytes,
+  APPROX_QUANTILES((bytesImg / 1024 / 1024 / 1024) * kw_per_GB * global_grid_intensity)[OFFSET(percentile * 10)] AS img_emissions,
+  APPROX_QUANTILES(bytesOther / 1024, 1000)[OFFSET(percentile * 10)] AS other_kbytes,
+  APPROX_QUANTILES((bytesOther / 1024 / 1024 / 1024) * kw_per_GB * global_grid_intensity)[OFFSET(percentile * 10)] AS other_emissions,
+  APPROX_QUANTILES(bytesHtmlDoc / 1024, 1000)[OFFSET(percentile * 10)] AS html_doc_kbytes,
+  APPROX_QUANTILES((bytesHtmlDoc / 1024 / 1024 / 1024) * kw_per_GB * global_grid_intensity)[OFFSET(percentile * 10)] AS html_doc_emissions,
+  APPROX_QUANTILES(bytesFont / 1024, 1000)[OFFSET(percentile * 10)] AS font_kbytes,
+  APPROX_QUANTILES((bytesFont / 1024 / 1024 / 1024) * kw_per_GB * global_grid_intensity)[OFFSET(percentile * 10)] AS font_emissions
+FROM
+  `httparchive.summary_pages.2022_06_01_*`,
+  UNNEST([10, 25, 50, 75, 90, 100]) AS percentile
```

Comment:
@tunetheweb how would I go about adding the Median Lighthouse Performance score to this query? I tried using a JOIN (below) but the results that were coming back in BigQuery didn't look right.

```JOIN (
    SELECT DISTINCT
    url,
    percentile,
    JSON_EXTRACT(report, '$.categories') AS categories
  FROM
    `httparchive.sample_data.lighthouse_*`,
    UNNEST([10, 25, 50, 75, 90, 100]) AS percentile
  )
USING
  (
    url,
    percentile
  )
```

```
APPROX_QUANTILES(CAST(JSON_VALUE(categories, '$.performance.score') AS NUMERIC), 1000)[OFFSET(percentile * 10)] AS median_performance,
```

Reply:
Something like this will work:

```sql
#standardSQL
# The distribution of page weight by resource type and client.

# Declare variables to calculate the carbon emissions of one byte
# Source: https://sustainablewebdesign.org/calculating-digital-emissions/
# The implementation below does not make the assumptions about returning visitors or caching that are present in the Sustainable Web Design model.

DECLARE kw_per_GB NUMERIC DEFAULT 0.81;
DECLARE global_grid_intensity NUMERIC DEFAULT 442;


SELECT
  percentile,
  _TABLE_SUFFIX AS client,
  APPROX_QUANTILES(bytesTotal / 1024, 1000)[OFFSET(percentile * 10)] AS total_kbytes,
  APPROX_QUANTILES((bytesTotal / 1024 / 1024 / 1024) * kw_per_GB * global_grid_intensity, 1000)[OFFSET(percentile * 10)] AS total_emissions,
  APPROX_QUANTILES(bytesHtml / 1024, 1000)[OFFSET(percentile * 10)] AS html_kbytes,
  APPROX_QUANTILES((bytesHtml / 1024 / 1024 / 1024) * kw_per_GB * global_grid_intensity, 1000)[OFFSET(percentile * 10)] AS html_emissions,
  APPROX_QUANTILES(bytesJS / 1024, 1000)[OFFSET(percentile * 10)] AS js_kbytes,
  APPROX_QUANTILES((bytesJS / 1024 / 1024 / 1024) * kw_per_GB * global_grid_intensity, 1000)[OFFSET(percentile * 10)] AS js_emissions,
  APPROX_QUANTILES(bytesCSS / 1024, 1000)[OFFSET(percentile * 10)] AS css_kbytes,
  APPROX_QUANTILES((bytesCSS / 1024 / 1024 / 1024) * kw_per_GB * global_grid_intensity, 1000)[OFFSET(percentile * 10)] AS css_emissions,
  APPROX_QUANTILES(bytesImg / 1024, 1000)[OFFSET(percentile * 10)] AS img_kbytes,
  APPROX_QUANTILES((bytesImg / 1024 / 1024 / 1024) * kw_per_GB * global_grid_intensity, 1000)[OFFSET(percentile * 10)] AS img_emissions,
  APPROX_QUANTILES(bytesOther / 1024, 1000)[OFFSET(percentile * 10)] AS other_kbytes,
  APPROX_QUANTILES((bytesOther / 1024 / 1024 / 1024) * kw_per_GB * global_grid_intensity, 1000)[OFFSET(percentile * 10)] AS other_emissions,
  APPROX_QUANTILES(bytesHtmlDoc / 1024, 1000)[OFFSET(percentile * 10)] AS html_doc_kbytes,
  APPROX_QUANTILES((bytesHtmlDoc / 1024 / 1024 / 1024) * kw_per_GB * global_grid_intensity, 1000)[OFFSET(percentile * 10)] AS html_doc_emissions,
  APPROX_QUANTILES(bytesFont / 1024, 1000)[OFFSET(percentile * 10)] AS font_kbytes,
  APPROX_QUANTILES((bytesFont / 1024 / 1024 / 1024) * kw_per_GB * global_grid_intensity, 1000)[OFFSET(percentile * 10)] AS font_emissions,
  APPROX_QUANTILES(score, 1000)[OFFSET(percentile * 10)] AS score
FROM
  `httparchive.summary_pages.2022_06_01_*`
JOIN
  (
    SELECT DISTINCT
      _TABLE_SUFFIX,
      url,
      JSON_EXTRACT(report, '$.categories.performance.score') AS score
    FROM
      `httparchive.lighthouse.2022_06_01_*`
  )
USING
  (url,_TABLE_SUFFIX),
  UNNEST([10, 25, 50, 75, 90, 100]) AS percentile
GROUP BY
  percentile,
  client
ORDER BY
  client,
  percentile
```

However I'm not sure what exactly you're expecting to get out of this? All the quantiles will be independent. So a site with low total kbytes could also have high css kbytes or vice versa. You don't have a common scale. And it will be the same with the Lighthouse scores. All it will tell you is 10% of sites have a lighthouse score less than or equal to 0.21, and 10 percent of sites have a total weight less than 528 KB, but it won't tell you if those are THE SAME sites.

Also we generally try to use the real-life CrUX data, rather than Lighthouse performance scores as a more accurate reflection of whether the site is performant for it's users.